torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -437,6 +437,10 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Tests if the UCC fallback path is available
   bool isUCCAvailable() const;
 
+  // Provides an API to abort the ProcessGroup (similar to ncclCommAbort)
+  // instead of relying on ProcessGroupNCCL destructor.
+  void abort(c10::optional<std::string> abortReason = c10::nullopt);
```

Comment:
wondering if we should directly call NCCL abort in the `~ProcessGroupNCCL` destructor as I feel `abort` should be part of the destruction process?

Providing an API specific to ProcessGroupNCCL would push user to use the low level API like NCCL pg directly (i.e. to use this feature), where given the dispatching collective that @H-Huang works on, I guess we should recommend high level API to user?

Reply:
> wondering if we should directly call NCCL abort in the ~ProcessGroupNCCL destructor as I feel abort should be part of the destruction process?

Thanks for the review @wanchaol :) Yes, abort is being called in the destructor of ProcessGroupNCCL on line 805 in ProcessGroupNCCL.cpp.

> Providing an API specific to ProcessGroupNCCL would push user to use the low level API like NCCL pg directly (i.e. to use this feature), where given the dispatching collective that @H-Huang works on, I guess we should recommend high level API to user?

I did think about adding a higher level API at the ProcessGroup level itself as well, but that might need additional thought since we would force all ProcessGroups to implement this new API. Happy to make changes here based on recommendations.

Reply:
Oh I see, yeah making it a ProcessGroup general API might need some more disscusion. I'm wondering if we don't expose the API to python, but rather call `del nccl_pg` in python to trigger the ProcessGroupNCCL destructor (or first remove the pg from the map), would it achieve similar purpose?

Is it because in some cases, since c10d maintains the map of pg where `del` sometimes can't trigger the process group destruction directly? If that's the case, curious what would be the case that it does not have access to the global pg map that c10d maintains?

Reply:
So `del nccl_pg` is just deleting your local reference and just decrements the ref count of the python object (it doesn't guarantee the destructor would be called). It would only trigger the ProcessGroupNCCL destructor once the ref count reaches 0. One example, is in my unit test itself where we pass in the process group to DDP and also into the Reducer in C++. Until the DDP and Reducer objects are not destroyed, the ref count for the object would not hit 0 and the object destructor would not be called.

In general it is pretty fragile to rely on destroying all references of the ProcessGroup since there could be several references to it throughout the codebase outside of PyTorch and c10d itself.

Reply:
I see, yeah that make sense, maybe we can discuss later about if we want to make this "force abort" as part of the process group protocol, starting with a private API looks good.